src/NovelRT/Ecs/Input/InputSystem.cpp

```diff
@@ -0,0 +1,91 @@
+                             std::shared_ptr<PluginManagement::IInputPluginProvider> inputProvider)
+    {
+        _actionMap = std::map<NovelRT::Atom, std::string>();
+        _firstUpdate = true;
+        _device = inputProvider->GetInputService();
+        _device->Initialise(windowingProvider->GetWindowingDevice()->GetHandle());
+    }
+
+    void InputSystem::Update(Timing::Timestamp delta, Ecs::Catalogue catalogue)
+    {
+        if (_firstUpdate)
+        {
+            auto actions = _device->GetAllMappings();
+            for (auto&& act : actions)
+            {
+                auto entityActionId = catalogue.CreateEntity();
+                _actionMap.insert(std::pair<NovelRT::Atom, std::string>(entityActionId, act.actionName));
+                _logger.logDebug("Input Mapped: \"{}\" to {}", act.actionName, act.pairedKey.GetKeyName());
+            }
+            _firstUpdate = false;
+        }
+
+        _device->Update(delta);
+
+        auto inputs = catalogue.GetComponentView<InputEventComponent>();
+
+        for (auto [entity, input] : inputs)
+        {
+            if (_device->GetKeyState(_actionMap.at(input.actionId)) == NovelRT::Input::KeyState::Idle)
+            {
+                inputs.RemoveComponent(entity);
+            }
+            else
+            {
+                auto event = InputEventComponent{input.actionId, _device->GetKeyState(_actionMap.at(input.actionId)),
+                                                 input.mousePositionX, input.mousePositionY};
+                inputs.PushComponentUpdateInstruction(entity, event);
+            }
+        }
+        inputs = catalogue.GetComponentView<InputEventComponent>();
+
+        for (auto action : _actionMap)
+        {
+            InputEventComponent in;
+            if (!inputs.TryGetComponent(action.first, in))
+            {
+                if (_device->GetKeyState(action.second) != NovelRT::Input::KeyState::Idle)
+                {
+                    auto mouse = _device->GetMousePosition();
+                    in = InputEventComponent{action.first, _device->GetKeyState(action.second), mouse.x, mouse.y};
+                    inputs.AddComponent(action.first, in);
+                }
+            }
+        }
+    }
+
+    void InputSystem::AddMapping(std::string name, std::string id)
+    {
+        unused(_device->AddInputAction(name, id));
+    }
+
+    void InputSystem::AddDefaultKBMMapping()
+    {
+        AddMapping("Up", "W");
+        AddMapping("Down", "S");
+        AddMapping("Left", "A");
+        AddMapping("Right", "D");
+        AddMapping("LeftClick", "LeftMouseButton");
+        AddMapping("A", "K");
+        AddMapping("B", "L");
+    }
+
+    std::string& InputSystem::GetMappingName(EntityId id)
+    {
+        return _actionMap.at(id);
+    }
+
+    InputSystem::~InputSystem() noexcept
+    {
+        unused(_actionMap.empty());
+        _device->TearDown();
```

Comment:
I don't understand why you are emptying the map.

Do we need to explicitly call TearDown here? Is it not handled by the dtor of the input device?

Reply:
Was just being explicit here - does this open up a possibility for UB?

Reply:
No, its just pointless if the `_device` is tearing down itself or if other things are using it. In the latter case that'll definitely cause issues.

As for the action map...there's no pointers there. It can delete itself.

Reply:
Removed this from InputSystem and the corresponding GLFW device as well.